long-ride-alerts/src/test/java/org/apache/flink/training/exercises/longrides/LongRidesTest.java

```diff
@@ -18,89 +18,150 @@
 import org.junit.Test;
 
 import java.time.Instant;
-import java.util.Collections;
-import java.util.List;
 
-import static org.junit.Assert.assertEquals;
+import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;
 
-public class LongRidesTest extends TaxiRideTestBase<TaxiRide> {
+public class LongRidesTest {
 
-    static final Testable JAVA_EXERCISE = () -> LongRidesExercise.main(new String[] {});
+    private static final int PARALLELISM = 2;
 
-    private static final Instant BEGINNING = Instant.parse("2020-01-01T12:00:00.00Z");
+    /** This isn't necessary, but speeds up the tests. */
+    @ClassRule
+    public static MiniClusterWithClientResource flinkCluster =
+            new MiniClusterWithClientResource(
+                    new MiniClusterResourceConfiguration.Builder()
+                            .setNumberSlotsPerTaskManager(PARALLELISM)
+                            .setNumberTaskManagers(1)
+                            .build());
+
+    public static final Instant BEGINNING = Instant.parse("2020-01-01T12:00:00.00Z");
+    public static final Instant ONE_MINUTE_LATER = BEGINNING.plusSeconds(60);
+    public static final Instant THREE_HOURS_LATER = BEGINNING.plusSeconds(180 * 60);
 
     @Test
     public void shortRide() throws Exception {
-        Instant oneMinLater = BEGINNING.plusSeconds(60);
+
         TaxiRide rideStarted = startRide(1, BEGINNING);
-        TaxiRide endedOneMinLater = endRide(rideStarted, oneMinLater);
-        Long markOneMinLater = oneMinLater.toEpochMilli();
+        TaxiRide endedOneMinLater = endRide(rideStarted, ONE_MINUTE_LATER);
 
-        TestRideSource source = new TestRideSource(rideStarted, endedOneMinLater, markOneMinLater);
-        assert (results(source).isEmpty());
+        ParallelTestSource<TaxiRide> source =
+                new ParallelTestSource<>(new PartitionByRideId(), rideStarted, endedOneMinLater);
+        TestSink<Long> sink = new TestSink<Long>();
+
+        longRidesPipeline().execute(source, sink);
+        assertThat(sink.results()).isEmpty();
     }
 
     @Test
     public void outOfOrder() throws Exception {
-        Instant oneMinLater = BEGINNING.plusSeconds(60);
         TaxiRide rideStarted = startRide(1, BEGINNING);
-        TaxiRide endedOneMinLater = endRide(rideStarted, oneMinLater);
-        Long markOneMinLater = oneMinLater.toEpochMilli();
+        TaxiRide endedOneMinLater = endRide(rideStarted, ONE_MINUTE_LATER);
+
+        ParallelTestSource<TaxiRide> source =
+                new ParallelTestSource<>(new PartitionByRideId(), endedOneMinLater, rideStarted);
+        TestSink<Long> sink = new TestSink<Long>();
 
-        TestRideSource source = new TestRideSource(endedOneMinLater, rideStarted, markOneMinLater);
-        assert (results(source).isEmpty());
+        longRidesPipeline().execute(source, sink);
+        assertThat(sink.results()).isEmpty();
     }
 
     @Test
     public void noStartShort() throws Exception {
```

Comment:
Is this test more of a UnitTest?

---

In general, the line may be difficult to draw, but how about this:
- an integration test focuses on the whole pipeline (so here: watermarks + keyBy + MatchFunction) and any interplay between these
- a unit test focuses on the logic of the MatchFunction alone

For this, we could actually also remove the watermarking code from the exercise and add this to the work to be done (unless you think, it may be too much to ask for a first exercise)